File: src/test/java/cube8540/oauth/authentication/credentials/oauth/token/infra/TokenExceptionTranslatorTest.java
package cube8540.oauth.authentication.credentials.oauth.token.infra;

import cube8540.oauth.authentication.credentials.oauth.token.domain.exception.TokenAccessDeniedException;
import cube8540.oauth.authentication.credentials.oauth.token.domain.exception.TokenNotFoundException;
import cube8540.oauth.authentication.error.message.ErrorMessage;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;

@DisplayName("토큰 에러 변환기 테스트")
class TokenExceptionTranslatorTest {

    private TokenExceptionTranslator translator;

    @BeforeEach
    void setup() {
        this.translator = new TokenExceptionTranslator();
    }

    @Test
    @DisplayName("TokenNotFoundException 변환")
    void translateTokenNotFoundException() {
        TokenNotFoundException exception = mock(TokenNotFoundException.class);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(exception);
        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
    }

    @Test
    @DisplayName("TokenAccessDeniedException 변환")
    void translateTokenAccessDeniedException() {
        TokenAccessDeniedException exception = mock(TokenAccessDeniedException.class);

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(exception);
        assertEquals(HttpStatus.FORBIDDEN, response.getStatusCode());
    }

    @Test
    @DisplayName("알 수 없는 예외 변환")
    void translateUnknownException() {
        Exception exception = new Exception();

        ResponseEntity<ErrorMessage<Object>> response = translator.translate(exception);
        assertEquals(HttpStatus.INTERNAL_SERVER_ERROR, response.getStatusCode());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/token/infra/TokenExceptionTranslatorTest.kt
package cube8540.oauth.authentication.oauth.token.infra

import cube8540.oauth.authentication.oauth.token.domain.TokenAccessDeniedException
import cube8540.oauth.authentication.oauth.token.domain.TokenNotFoundException
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.springframework.http.HttpStatus
import java.util.stream.Stream

class TokenExceptionTranslatorTest {

    private val translator: TokenExceptionTranslator = TokenExceptionTranslator()

    @ParameterizedTest
    @MethodSource(value = ["tokenAccessDeniedException"])
    fun `translate token access denied exception`(exception: TokenAccessDeniedException) {
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.FORBIDDEN)
    }

    @ParameterizedTest
    @MethodSource(value = ["tokenNotFoundException"])
    fun `translate token not found exception`(exception: TokenNotFoundException) {
        val result = translator.translate(exception)

        assertThat(result.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }

    private fun tokenAccessDeniedException() = Stream.of(
        Arguments.of(TokenAccessDeniedException.denied("denied"))
    )

    private fun tokenNotFoundException() = Stream.of(
        Arguments.of(TokenNotFoundException.instance("not found"))
    )
}